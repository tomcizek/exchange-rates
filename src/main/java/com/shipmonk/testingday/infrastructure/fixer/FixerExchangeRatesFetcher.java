package com.shipmonk.testingday.infrastructure.fixer;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import org.json.JSONObject;
import org.springframework.stereotype.Service;

@Service
public class FixerExchangeRatesFetcher {

    private final HttpClient client;

    public FixerExchangeRatesFetcher(HttpClient client) {
        this.client = client;
    }

    public JSONObject fetchRates(String url) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(url))
            .GET()
            .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        return new JSONObject(response.body());
    }

}
