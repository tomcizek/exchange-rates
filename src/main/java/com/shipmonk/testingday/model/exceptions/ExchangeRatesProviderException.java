package com.shipmonk.testingday.model.exceptions;

public class ExchangeRatesProviderException extends Exception {


    public ExchangeRatesProviderException(
        String message
    ) {
        super(message);
    }

    public ExchangeRatesProviderException(
        String message,
        Throwable cause
    ) {
        super(message, cause);
    }

    public static ExchangeRatesProviderException becauseNoConversionSymbolsProvided() {
        return new ExchangeRatesProviderException(
            "No symbols for currency conversion provided."
        );
    }

    public static ExchangeRatesProviderException becauseFetchingFailed(Exception e) {
        return new ExchangeRatesProviderException(
            "Fetching exchange rates failed.",
            e
        );
    }

    public static ExchangeRatesProviderException becauseParsingResultJSONFailed(Exception e) {
        return new ExchangeRatesProviderException(
            "Parsing JSON result of exchange rates failed.",
            e
        );
    }
}
