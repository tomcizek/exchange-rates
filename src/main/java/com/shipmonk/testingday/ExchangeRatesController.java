package com.shipmonk.testingday;

import com.shipmonk.testingday.model.ExchangeRates;
import com.shipmonk.testingday.model.ExchangeRatesProvider;
import com.shipmonk.testingday.model.exceptions.ExchangeRatesProviderException;
import java.time.LocalDate;
import java.time.format.DateTimeParseException;
import java.util.Arrays;
import java.util.List;
import javax.money.CurrencyUnit;
import javax.money.Monetary;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping(
    path = "/api/v1/rates"
)
public class ExchangeRatesController {

    private ExchangeRatesProvider exchangeRatesProvider;

    @Value("${exchange.rates.currency.target.defaults}")
    private String defaultTargetCurrencies;

    public ExchangeRatesController(ExchangeRatesProvider exchangeRatesProvider) {
        this.exchangeRatesProvider = exchangeRatesProvider;
    }

    @RequestMapping(method = RequestMethod.GET, path = "/{day}")
    public ResponseEntity<Object> getRates(@PathVariable("day") String day) {

        LocalDate localDate;
        try {
            localDate = LocalDate.parse(day);
        } catch (DateTimeParseException e) {
            return new ResponseEntity<>(
                String.format("Invalid date format: '%s'", day),
                HttpStatus.BAD_REQUEST
            );
        }

        List<CurrencyUnit> defaultTargetCurrencies;
        try {
            defaultTargetCurrencies = Arrays.stream(this.defaultTargetCurrencies.split(","))
                .map(Monetary::getCurrency)
                .toList();
        } catch (Exception e) {
            // TODO log exception, bad configuration
            return new ResponseEntity<>(
                "Invalid currency symbols", HttpStatus.INTERNAL_SERVER_ERROR);
        }

        try {
            ExchangeRates rates = exchangeRatesProvider.getRates(
                localDate, defaultTargetCurrencies);
            return new ResponseEntity<>(rates, HttpStatus.OK);
        } catch (ExchangeRatesProviderException e) {
            return new ResponseEntity<>(e.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

}
