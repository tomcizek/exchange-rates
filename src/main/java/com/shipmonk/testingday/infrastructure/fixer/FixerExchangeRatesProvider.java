package com.shipmonk.testingday.infrastructure.fixer;

import com.shipmonk.testingday.model.ExchangeRates;
import com.shipmonk.testingday.model.ExchangeRatesProvider;
import com.shipmonk.testingday.model.exceptions.ExchangeRatesProviderException;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import javax.money.CurrencyUnit;
import org.json.JSONObject;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Service
public class FixerExchangeRatesProvider implements ExchangeRatesProvider {

    private final FixerExchangeRatesFetcher fixerExchangeRatesFetcher;

    @Value("${fixer.api.key}")
    private String apiKey;

    @Value("${fixer.api.base.url}")
    private String baseURL;

    @Value("${fixer.api.currency.base.default}")
    private String defaultBaseCurrency;

    public FixerExchangeRatesProvider(FixerExchangeRatesFetcher fixerExchangeRatesFetcher) {
        this.fixerExchangeRatesFetcher = fixerExchangeRatesFetcher;
    }

    @Override
    public ExchangeRates getRates(
        LocalDate day,
        List<CurrencyUnit> symbols
    ) throws ExchangeRatesProviderException {
        if (symbols.isEmpty()) {
            throw ExchangeRatesProviderException.becauseNoConversionSymbolsProvided();
        }

        JSONObject jsonResponse;
        var url = buildUrl(day, symbols);
        try {
            jsonResponse = this.fixerExchangeRatesFetcher.fetchRates(url);
        } catch (Exception e) {
            throw ExchangeRatesProviderException.becauseFetchingFailed(e);
        }

        try {
            return parseJson(jsonResponse);
        } catch (Exception e) {
            throw ExchangeRatesProviderException.becauseParsingResultJSONFailed(e);
        }
    }

    private String buildUrl(LocalDate day, List<CurrencyUnit> symbols) {
        String date = day.toString();

        return this.baseURL + date +
            "?access_key=" + this.apiKey +
//            "&base=" + baseCurrency +
            "&symbols=" + String.join(",", symbols.stream()
            .map(CurrencyUnit::getCurrencyCode)
            .toList()
        );
    }

    private ExchangeRates parseJson(JSONObject jsonObject) {
        boolean success = jsonObject.getBoolean("success");
        String base = jsonObject.getString("base");
        LocalDate date = LocalDate.parse(jsonObject.getString("date"));
        JSONObject ratesObject = jsonObject.getJSONObject("rates");

        Map<String, BigDecimal> rates = new HashMap<>();
        Iterator<String> keys = ratesObject.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            rates.put(key, BigDecimal.valueOf(ratesObject.getDouble(key)));
        }

        return new ExchangeRates(
            success,
            base,
            date,
            rates
        );
    }
}
