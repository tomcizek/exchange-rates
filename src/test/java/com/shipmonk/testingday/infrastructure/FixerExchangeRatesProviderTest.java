package com.shipmonk.testingday.infrastructure;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.shipmonk.testingday.ApprovalTest;
import com.shipmonk.testingday.infrastructure.fixer.FixerExchangeRatesProvider;
import com.shipmonk.testingday.model.exceptions.ExchangeRatesProviderException;
import java.time.LocalDate;
import java.util.List;
import javax.money.Monetary;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

@SpringBootTest
class FixerExchangeRatesProviderTest extends ApprovalTest {

    @Autowired
    private FixerExchangeRatesProvider fixerExchangeRatesProvider;

    @Test
    public void itShouldFailWhenNoConversionSymbolsProvided() {
        Executable getRatesCall = () -> fixerExchangeRatesProvider.getRates(
            LocalDate.of(2023, 6, 6),
            List.of()
        );

        var exception = assertThrows(
            ExchangeRatesProviderException.class,
            getRatesCall
        );

        assertEquals("No symbols for currency conversion provided.", exception.getMessage());
    }

    @Test
    public void itShouldReturnExchangeRates() throws ExchangeRatesProviderException {
        var exchangeRates = fixerExchangeRatesProvider.getRates(
            LocalDate.of(2023, 6, 6),
            List.of(
                Monetary.getCurrency("EUR"),
                Monetary.getCurrency("CZK")
            )
        );

        this.thenExpectedResult(exchangeRates);
    }
}
