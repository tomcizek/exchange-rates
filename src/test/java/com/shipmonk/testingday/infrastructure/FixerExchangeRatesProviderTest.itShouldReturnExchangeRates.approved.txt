{
  "success" : true,
  "base" : "EUR",
  "date" : "2023-06-06",
  "rates" : {
    "EUR" : 1.0,
    "CZK" : 23.578329
  }
}
